### PR TITLE
feat:[SSCA-400]: Added spec for required OPA Apis

### PR DESCRIPTION
**Pull request by arpit-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51826)

### CODE REVIEW

**Comment by arpit-harness** [comment]
trigger uts

**Review by deepharness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51826)* status: 1 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`ssca-manager/contracts/openapi/v1/openapi.yaml` line 43 at r1](https://reviewable.io/reviews/harness/harness-core/51826#-NcLSr1cDzXUv6Xdgd1h:-NcLSr1cDzXUv6Xdgd1i:bz4fqdy) ([raw file](https://github.com/harness/harness-core/blob/a9d9d339064ca3804a5e3814c9b0c30eec45154d/ssca-manager/contracts/openapi/v1/openapi.yaml#L43)):*
> ```YAML
>       tags:
>         - sbomProcessor
>   '/v1/org/{org}/project/{project}/normalized_sbom_component/list':
> ```

We need component based on orchestrationId , not all the components of whole project

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 1]
```diff
@@ -40,6 +40,100 @@ paths:
         $ref: '#/components/requestBodies/SbomProcessRequestBody'
       tags:
         - sbomProcessor
+  '/v1/org/{org}/project/{project}/normalized_sbom_component/list':
```
Comment (mankrit-singh) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:43]: normalized-sbom-component (hyphenated)
also can you remove the "/list"? 
If it's a GET call without resource identifier it is understood as List

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 2]
```diff
@@ -40,6 +40,100 @@ paths:
...
+          type: string
+        description: Harness Organization Id
+      - in: path
+        required: true
+        schema:
+          type: string
+        description: Harness Project Id
+        name: project
+    get:
+      summary: ''
```
Comment (mankrit-singh) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:58]: please add a summary

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 3]
```diff
@@ -40,6 +40,100 @@ paths:
...
+        '200':
+          description: OK
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/NormalizedSbomComponentDTO'
+          headers:
+            header-0:
```
Comment (mankrit-singh) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:70]: what is this used for?

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 4]
```diff
@@ -40,6 +40,100 @@ paths:
...
+              schema:
+                type: string
+      description: Gives a paginated list of normalized SBOM Components.
+      tags:
+        - normalizeSbom
+      parameters:
+        - $ref: '#/components/parameters/PageNumberParam'
+        - $ref: '#/components/parameters/PageSizeParam'
+  '/v1/org/{org}/project/{project}/enforcement_result':
```
Comment (mankrit-singh) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:84]: hyphenated here too

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 5]
```diff
@@ -40,6 +40,100 @@ paths:
...
+          type: string
+        description: Harness Organization Id
+      - in: path
+        required: true
+        schema:
+          type: string
+        description: Harness Project Id
+        name: project
+    post:
+      summary: ''
```
Comment (mankrit-singh) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:99]: summary

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 6]
```diff
@@ -40,6 +40,100 @@ paths:
...
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/EnforcementResultDTO'
+        description: Enforcement Result to be saved.
+      tags:
+        - enforcementResult
+      description: save enforcement result details.
+  '/v1/org/{org}/project/{project}/enforcement_summary':
```
Comment (mankrit-singh) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:113]: hyphenated here too

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 7]
```diff
@@ -236,6 +330,143 @@ components:
           $ref: '#/components/schemas/Artifact'
         policy_file_id:
           type: string
+    NormalizedSbomComponentDTO:
+      title: NormalizedSbomComponentDTO
+      x-stoplight:
+        id: f4842caddn13n
+      type: object
+      properties:
+        orchestrationId:
```
Comment (mankrit-singh) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:344]: underscore casing for all these parameters eg.  orchestration_id:

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 8]
```diff
@@ -236,6 +330,143 @@ components:
...
+          type: string
+        artifactId:
+          type: string
+        artifactName:
+          type: string
+        tags:
+          type: array
+          items:
+            type: string
+        createdOn:
```
Comment (mankrit-singh) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:358]: [created](https://harness.atlassian.net/wiki/spaces/PLATFORM/pages/21203550744/Summary+of+API+Redesign+Changes#API-Standards)

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 9]
```diff
@@ -40,6 +40,105 @@ paths:
...
       tags:
         - sbomProcessor
+  '/v1/org/{org}/project/{project}/normalized-sbom-component':
+    parameters:
+      - name: org
+        in: path
+        required: true
+        schema:
+          type: string
+        description: Harness Organization Id
```
Comment (deepakbolangady) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:50]: Please change globally.
```suggestion
        description: Harness organization ID
```

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 10]
```diff
@@ -40,6 +40,105 @@ paths:
...
+      - in: path
+        required: true
+        schema:
+          type: string
+        description: Harness Project Id
```
Comment (deepakbolangady) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:55]: Please change globally.
```suggestion
        description: Harness project ID
```

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 11]
```diff
@@ -40,6 +40,105 @@ paths:
...
+        name: project
+    get:
+      summary: Get Normalised SBOM Components
```
Comment (deepakbolangady) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:58]: Use sentence case globally. Some descriptions are also starting with lowercase.

For example (but apply everywwhere):
```suggestion
      summary: Get normalised SBOM components
```

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 12]
```diff
@@ -40,6 +40,105 @@ paths:
...
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/NormalizedSbomComponentDTO'
+          headers:
+            header-0:
+              schema:
+                type: string
+      description: Gives a paginated list of normalized SBOM Components.
```
Comment (deepakbolangady) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:73]:
```suggestion
      description: Gives a paginated list of normalized SBOM components.
```

File: ssca-manager/contracts/openapi/v1/openapi.yaml [hunk 13]
```diff
@@ -40,6 +40,105 @@ paths:
...
+      tags:
+        - normalizeSbom
+      parameters:
+        - $ref: '#/components/parameters/PageNumberParam'
+        - $ref: '#/components/parameters/PageSizeParam'
+        - schema:
+            type: string
+          in: query
+          name: orchestrationId
+          description: Required Orchestration step ID
```
Comment (deepakbolangady) [inline ssca-manager/contracts/openapi/v1/openapi.yaml:83]:
```suggestion
          description: Required orchestration step ID
```

**Review by deepakbolangady — DISMISSED** [review]
Approving with some feedback.

**Comment by arpit-harness** [comment]
trigger uts

**Review by deepharness — CHANGES_REQUESTED** [review]
> Reviewed 24 of 24 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51826)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[`ssca-manager/service/src/main/java/io/harness/repositories/SBOMComponentRepo.java` line 24 at r3](https://reviewable.io/reviews/harness/harness-core/51826#-NcVWXii9UYIgao1tNe-:-NcVWXii9UYIgao1tNe0:b-mhkugx) ([raw file](https://github.com/harness/harness-core/blob/dd83615bd3a13bfa9b8ce6e6343fb98dbed6d707/ssca-manager/service/src/main/java/io/harness/repositories/SBOMComponentRepo.java#L24)):*
> ```Java
> public interface SBOMComponentRepo
>     extends PagingAndSortingRepository<NormalizedSBOMComponentEntity, String>, SBOMComponentRepoCustom {
>   Page<NormalizedSBOMComponentEntity> findByAccountIdAndOrgIdentifierAndProjectIdentifierAndOrchestrationId(
> ```

Too long for name , findBy is good enough. Also just fyi when implementing this , orchestrationId is optional , if orchestrationId is null or empty , then we should return latest components from last execution.
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/NormalizeSbomServiceImpl.java` line 21 at r3](https://reviewable.io/reviews/harness/harness-core/51826#-NcVY_jY9xAIU-caLrKS:-NcVY_jY9xAIU-caLrKT:b-j3dawf) ([raw file](https://github.com/harness/harness-core/blob/dd83615bd3a13bfa9b8ce6e6343fb98dbed6d707/ssca-manager/service/src/main/java/io/harness/ssca/services/NormalizeSbomServiceImpl.java#L21)):*
> ```Java
> import org.springframework.data.domain.Pageable;
> 
> public class NormalizeSbomServiceImpl implements NormalizeSbomService {
> ```

We don't need this , its doing a middle ware job. Instead create a middleware which can parse the paginated response.
Here are few examples 
https://www.baeldung.com/rest-api-pagination-in-spring
https://www.geeksforgeeks.org/spring-rest-pagination/
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/NormalizeSbomServiceImpl.java` line 31 at r3](https://reviewable.io/reviews/harness/harness-core/51826#-NcVXtUmAOUjoLNpMN46:-NcVXtUmAOUjoLNpMN47:bnez666) ([raw file](https://github.com/harness/harness-core/blob/dd83615bd3a13bfa9b8ce6e6343fb98dbed6d707/ssca-manager/service/src/main/java/io/harness/ssca/services/NormalizeSbomServiceImpl.java#L31)):*
> ```Java
>         sbomComponentRepo.findByAccountIdAndOrgIdentifierAndProjectIdentifierAndOrchestrationId(
>             accountId, orgIdentifier, projectIdentifier, orchestrationId, pageRequest);
>     Page<NormalizedSbomComponentDTO> result = entities.map(entity -> NormalizeSbomComponentTransformer.toDTO(entity));
> ```

this is a typical middle ware job , this will be required for any paginated api. We should create a separate interface called PaginatedResponse<T> for generating paginated response from Page<T>
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/utils/transformers/EnforcementResultTransformer.java` line 13 at r3](https://reviewable.io/reviews/harness/harness-core/51826#-NcVZWI_1OZk4jWP1ode:-NcVZWI_1OZk4jWP1odf:b-lgeyqa) ([raw file](https://github.com/harness/harness-core/blob/dd83615bd3a13bfa9b8ce6e6343fb98dbed6d707/ssca-manager/service/src/main/java/io/harness/ssca/utils/transformers/EnforcementResultTransformer.java#L13)):*
> ```Java
> import io.harness.ssca.entities.EnforcementResultEntity;
> 
> public class EnforcementResultTransformer {
> ```

We should use generic data mappers like http://modelmapper.org/ . Or else for every new dto & entity we will keep writing new converters 
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/utils/transformers/EnforcementSummaryTransformer.java` line 16 at r3](https://reviewable.io/reviews/harness/harness-core/51826#-NcVcAenBrFgZp0MImCh:-NcVcAenBrFgZp0MImCi:b-lgeyqa) ([raw file](https://github.com/harness/harness-core/blob/dd83615bd3a13bfa9b8ce6e6343fb98dbed6d707/ssca-manager/service/src/main/java/io/harness/ssca/utils/transformers/EnforcementSummaryTransformer.java#L16)):*
> ```Java
> import java.math.BigDecimal;
> 
> public class EnforcementSummaryTransformer {
> ```

We should use generic data mappers like http://modelmapper.org/ . Or else for every new dto & entity we will keep writing new converters 
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/utils/transformers/NormalizeSbomComponentTransformer.java` line 16 at r3](https://reviewable.io/reviews/harness/harness-core/51826#-NcVcCyd85SJVZqLWlOL:-NcVcCyd85SJVZqLWlOM:b-lgeyqa) ([raw file](https://github.com/harness/harness-core/blob/dd83615bd3a13bfa9b8ce6e6343fb98dbed6d707/ssca-manager/service/src/main/java/io/harness/ssca/utils/transformers/NormalizeSbomComponentTransformer.java#L16)):*
> ```Java
> import java.time.Instant;
> 
> public class NormalizeSbomComponentTransformer {
> ```

We should use generic data mappers like http://modelmapper.org/ . Or else for every new dto & entity we will keep writing new converters

**Review by deepharness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51826)* status: 3 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`ssca-manager/service/src/main/java/io/harness/repositories/SBOMComponentRepo.java` line 24 at r4](https://reviewable.io/reviews/harness/harness-core/51826#-NcfDO4n1k7oebDTVH_V:-NcfDO4n1k7oebDTVH_W:b-1bhyia) ([raw file](https://github.com/harness/harness-core/blob/c04fa64c8f9da4414c09c67e7d320e9312fbabf2/ssca-manager/service/src/main/java/io/harness/repositories/SBOMComponentRepo.java#L24)):*
> ```Java
> public interface SBOMComponentRepo
>     extends PagingAndSortingRepository<NormalizedSBOMComponentEntity, String>, SBOMComponentRepoCustom {
>   Page<NormalizedSBOMComponentEntity> findByAccountIdAndOrgIdentifierAndProjectIdentifierAndOrchestrationId(
> ```

**Pull request closed without merging**